File: app/utils/dawproject_python/warp.py
from lxml import etree as ET


class Warp:
    def __init__(self, time, content_time):
        self.time = time
        self.content_time = content_time

    def to_xml(self):
        warp_elem = ET.Element("Warp")
        warp_elem.set("time", str(self.time))
        warp_elem.set("contentTime", str(self.content_time))
        return warp_elem

    @classmethod
    def from_xml(cls, element):
        time = float(element.get("time"))
        content_time = float(element.get("contentTime"))
        return cls(time, content_time)

File: app/utils/dawproject/timeline/warp.ts
import { DoubleAdapter } from "../doubleAdapter";
import type { IWarp } from "../types";
import { Utility } from "../utility";
import { XmlObject } from "../XmlObject";

export class Warp extends XmlObject implements IWarp {
  time: number;
  contentTime: number;

  constructor(time?: number, contentTime?: number) {
    super();
    // Make required fields optional for deserialization, fromXmlObject will set them
    this.time = time || 0; // Provide a default placeholder
    this.contentTime = contentTime || 0; // Provide a default placeholder
  }

  toXmlObject(): any {
    const obj: any = {
      Warp: {},
    };

    // add required attributes
    Utility.addAttribute(obj.Warp, "time", this, {
      required: true,
      adapter: DoubleAdapter.toXml,
    });
    Utility.addAttribute(obj.Warp, "contentTime", this, {
      required: true,
      adapter: DoubleAdapter.toXml,
    });

    return obj;
  }

  fromXmlObject(xmlObject: any): this {
    // validate and populate required attributes
    Utility.populateAttribute<number>(xmlObject, "time", this, {
      required: true,
      adapter: DoubleAdapter.fromXml,
    });
    Utility.populateAttribute<number>(xmlObject, "contentTime", this, {
      required: true,
      adapter: DoubleAdapter.fromXml,
    });

    return this;
  }
}
